### ISSUE TITLE
Support for solving a batch of problems

### ISSUE THREAD

**Issue by chrisyeh96** [cvxpy/cvxpy#2485]
## Background, motivation, and inspiration

I would love for cvxpy to support the ability to solve a batch of similar problems. Specifically, I am interested in the setting where the problem structure is fixed, and there is a batch of different problem parameters.

The motivation and inspiration for this feature request comes from cvxpylayers (which internally depends on diffcp), which supports solving a batch of cvxpy problems using a ThreadPool. See code here:
https://github.com/cvxgrp/diffcp/blob/cdef024c6b656814f92e3ca10f47ae4feaace1fb/diffcp/cone_program.py#L90-L94.

cvxpylayers allows users to define a problem in cvxpy, then pass in parameters as PyTorch `torch.Tensor` objects, which are allowed to have a batch dimension. In the following example code, a standard cvxpy problem is defined. All the `cp.Variable` and `cp.Parameter` objects are written for a single problem (no batch dimension). However, the parameters in PyTorch have a `batch_size` dimension, and the problem solution is returned with a batch dimension as well:

```python
import cvxpy as cp
from cvxpylayers.torch import CvxpyLayer
import torch

n, m = 2, 3
x = cp.Variable(n)
A = cp.Parameter((m, n))
b = cp.Parameter(m)
constraints = [x >= 0]
objective = cp.Minimize(0.5 * cp.pnorm(A @ x - b, p=1))
problem = cp.Problem(objective, constraints)
assert problem.is_dpp()

cvxpylayer = CvxpyLayer(problem, parameters=[A, b], variables=[x])

batch_size = 5
A_tch = torch.randn(batch_size, m, n, requires_grad=True)
b_tch = torch.randn(batch_size, m, requires_grad=True)

# solve the problem
solution, = cvxpylayer(A_tch, b_tch)

print(solution.shape)
```

Output:
```
torch.Size([5, 2])
```

More generally, I am motivated by the problem of wanting to solve a number of optimization problems as part of the forward inference pass of a neural network. cvxpylayers (via diffcp) always calculates the gradient as well, but at inference time, I do not need the gradient information. Therefore, I hope that cvxpy can directly solve a batch of problems.

## Possible API and design considerations

I see several possible APIs for supporting a batch of problem parameters, and the following ideas are by no means a comprehensive list.

**1) `cp.Parameter.batch_value`, `cp.Variable.batch_value`, and `prob.solve(batch=True)`**

This idea would essentially mimic the cvxpylayers approach, where the `cp.Problem` is written for a single instance of parameters, but the user can pass in a batch of parameter values. One possible way to do this is

```python
var = cp.Variable(n)
param = cp.Parameter(m, n)
param.batch_value = np.ones([batch_size, m, n])
...
prob.solve(batch=True)

# solution is stored in:
var.batch_value  # shape [batch_size, n]

# optimal value is stored as:
prob.batch_value  # shape [batch_size]
```

The main benefit of this approach is that the batch size does not need to specified at the time the problem is created. However, it likely necessitates a `batch_value` (or similar) attribute for `cp.Parameter` and `cp.Variable`.

**2) `cp.Parameter.set_value(..., batch=True)`**

One possible workaround to adding a `batch_value` attribute would be to add a new `set_value()` method onto `cp.Parameter`. This would allow keeping `param.value` and `var.value` as the only ways to access parameter and variable values, with the understanding that any call to `cp.Parameter.set_value(..., batch=True)` automatically adds a batch dimension to parameter and variable values.

```python
var = cp.Variable(n)
param = cp.Parameter(m, n)
param.set_value(np.ones([batch_size, m, n]), batch=True)
...
prob.solve()

# solution is stored in:
var.value  # shape [batch_size, n]

# optimal value is stored as:
prob.value  # shape [batch_size]
```

**3) `cp.Variable(None, n)` and `cp.Parameter(None, m, n)`**

A third option is to allow `None` as a placeholder for a batch dimension in `cp.Variable` and `cp.Parameter`.

```python
x = cp.Variable(None, n)
param = cp.Parameter(None, m, n)
param.value = np.ones([batch_size, m, n])
...
prob.solve()

# solution is stored in:
x.value  # shape [batch_size, n]

# optimal value is stored as:
prob.value  # shape [batch_size]
```

## Discussion

Personally, I think options (2) or (3) above would be the most natural for me.

Of course, there are plenty of other choices to be made, such as what `constr.violation()` should return, and how `prob.status` should be handled. Looking forward to brainstorming ideas to make this API intuitive and flexible.

**Comment by SteveDiamond** [comment]
Thanks for the detailed explanation! We are actively working on adding n-dimensional support to cvxpy. At that point a batch dimension could be handled as a normal extra dimension (e.g., `Variable(shape=(m,n,batch_size))`) I believe. If you construct a single problem that includes batching, it will be separable but that doesn't matter necessarily. There's no harm in solving it as one large problem.

**Comment by PTNobel** [comment]
>  you construct a single problem that includes batching, it will be separable but that doesn't matter necessarily. There's no harm in solving it as one large problem.

I think I disagree. Under Chris's proposal, we'd dispatch multiple threads in parallel for the batches. I don't think any of our solvers currently support recognizing separable problems and solving them in parallel.

**Comment by SteveDiamond** [comment]
But is that actually faster than solving one large problem with the parallel linear algebra that happens in the more sophisticated solvers? I haven't found it to be the case when I've looked into this before.

**Comment by SteveDiamond** [comment]
Since I posted the above it became more clear that in many practical situations a batched approach is indeed faster than processing one massive problem. So at this point I am in favor of supporting batching.
